File: VB/WebSite/App_Code/FileManagerAccessRuleHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports DevExpress.Web.ASPxFileManager

''' <summary>
''' Summary description for FileManagerAccessRuleHelper
''' </summary>
Public Class FileManagerAccessRuleHelper
	Private fm As ASPxFileManager
	Private compoundUserRole_Renamed As String

	Public ReadOnly Property CompoundUserRole() As String
		Get
			If compoundUserRole_Renamed Is Nothing Then
				compoundUserRole_Renamed = String.Join("|", System.Web.Security.Roles.GetRolesForUser())
			End If
			Return compoundUserRole_Renamed
		End Get
	End Property

	Public Sub New(ByVal fileManager As ASPxFileManager)
		Me.fm = fileManager
	End Sub
	Public Sub CreateRuleForCompoundRole()
		Dim userRoles() As String = System.Web.Security.Roles.GetRolesForUser()
		Dim roleAccessRules As New List(Of FileManagerAccessRuleBase)()
		Dim currentAccessRules As AccessRulesCollection = fm.SettingsPermissions.AccessRules

		For Each userRole As String In userRoles
			roleAccessRules.AddRange(currentAccessRules.FindAll(Function(r) String.Compare(r.Role, userRole, True) = 0))
		Next userRole

		CreateCompoundRoleRules(roleAccessRules)

	End Sub
	Private Sub CreateCompoundRoleRules(ByVal userRolesRuleList As List(Of FileManagerAccessRuleBase))
		For Each role As FileManagerAccessRuleBase In userRolesRuleList
			fm.SettingsPermissions.AccessRules.Add(CreateCompoundRoleRule(role))
		Next role
	End Sub

	Private Function CreateCompoundRoleRule(ByVal source As FileManagerAccessRuleBase) As FileManagerAccessRuleBase
		Dim rule As FileManagerAccessRuleBase
		If TypeOf source Is FileManagerFolderAccessRule Then
			rule = New FileManagerFolderAccessRule()
		Else
			rule = New FileManagerFileAccessRule()
		End If

		rule.Assign(source)
		rule.Role = CompoundUserRole
		Return rule
	End Function
End Class

File: CS/WebSite/App_Code/FileManagerAccessRuleHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using DevExpress.Web.ASPxFileManager;

/// <summary>
/// Summary description for FileManagerAccessRuleHelper
/// </summary>
public class FileManagerAccessRuleHelper
{
	ASPxFileManager fm;
	private string compoundUserRole;

	public string CompoundUserRole {
		get {
			if (compoundUserRole == null) {
				compoundUserRole = String.Join("|", System.Web.Security.Roles.GetRolesForUser());
			}
			return compoundUserRole;
		}
	}

	public FileManagerAccessRuleHelper(ASPxFileManager fileManager) {
		this.fm = fileManager;
	}
	public void CreateRuleForCompoundRole() {
		string[] userRoles = System.Web.Security.Roles.GetRolesForUser();
		List<FileManagerAccessRuleBase> roleAccessRules = new List<FileManagerAccessRuleBase>();
		AccessRulesCollection currentAccessRules = fm.SettingsPermissions.AccessRules;

		foreach (string userRole in userRoles)
			roleAccessRules.AddRange(currentAccessRules.FindAll(r => String.Compare(r.Role, userRole, true) == 0));

		CreateCompoundRoleRules(roleAccessRules);

	}
	private void CreateCompoundRoleRules( List<FileManagerAccessRuleBase> userRolesRuleList) {
		foreach (FileManagerAccessRuleBase role in userRolesRuleList) {
			fm.SettingsPermissions.AccessRules.Add(CreateCompoundRoleRule(role));
		}
	}

	private FileManagerAccessRuleBase CreateCompoundRoleRule(FileManagerAccessRuleBase source) {
		FileManagerAccessRuleBase rule;
		if (source is FileManagerFolderAccessRule)
			rule = new FileManagerFolderAccessRule();
		else
			rule = new FileManagerFileAccessRule();

		rule.Assign(source);
		rule.Role = CompoundUserRole;
		return rule;
	}
}